### PR TITLE
improve BCALM input handling & sorting; reduce memory usage ~25%

### LINKED ISSUE #283: sort bcalm unitigs using md5sum
one of the hacks we use to force determinism in sgc is to sort the unitigs into a specific order, since unitig order is not guaranteed by bcalm in multithread mode. this is currently very heavyweight because we load in all the sequences and hold them in memory. for big data sets, this is problematic.

(yes, I know, it's dumb, but it worked at the time ;)

I think we could probably just switch to sorting the unitigs using their md5sum or some other hash.

### LINKED ISSUE #299: potential bug during pendant removal
in the `bcalm_to_gxt` step, some nodes are removed & merged with others in order to simplify the graph. during the ensuing indexing of k-mers, it looks like the k-mers that belong to these removed nodes are not counted.

this is easy to see if you use a linear genome which has pendants on either end that get removed ;). basically the k-mers from the beginning and end of the contigs don't make it into the MPHF index.

this should have a (very) small overall effect on real metagenomes, though.

### LINKED ISSUE #301: Potential error in rule bcalm_catlas_input with SLURM & Auk
Hi again. While rerunning my spacegraphcats run that had failed to finish in the allotted wall time, I ran into the follow warning/error during the bcalm_catlas_input rule. `Auks API request failed : auks cred : credential expired` . The full log from the rerun is below. 
```
...sgc is updating input file data/RR/RR452Oct_Bulk_S26_abundtrim.fq.gz with outdir prefix /share/emersonlab/sorensen/spacegraphcats
...sgc is updating input file data/RR/HQRR.18.fa with outdir prefix /share/emersonlab/sorensen/spacegraphcats
...sgc is updating input file data/RR/HQRR.25.fa with outdir prefix /share/emersonlab/sorensen/spacegraphcats
...sgc is updating input file data/RR/LQRR.84.fa with outdir prefix /share/emersonlab/sorensen/spacegraphcats
...sgc is updating input file data/RR/LQRR.8.fa with outdir prefix /share/emersonlab/sorensen/spacegraphcats
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	bcalm_catlas_input
	1	build_catlas
	1	make_contigs_kmer_index
	1	search
	4

[Fri Oct  2 04:03:18 2020]
rule bcalm_catlas_input:
    input: RR_test/bcalm.RR_test.k21.unitigs.fa
    output: RR_test_k21_r1/cdbg.gxt, RR_test_k21_r1/contigs.fa.gz, RR_test_k21_r1/contigs.fa.gz.info.csv
    jobid: 3

Auks API request failed : auks cred : credential expired
slurmstepd: error: *** JOB 27710792 ON rafter-0 CANCELLED AT 2020-10-04T05:03:10 DUE TO TIME LIMIT ***
slurmstepd: error: *** JOB 27710792 STEPD TERMINATED ON rafter-0 AT 2020-10-04T05:04:42 DUE TO JOB NOT ENDING WITH SIGNALS ***
```
The job again didn't finish, I think it somehow "hung" during this rule. Running `ls -la RR_test_k21_r1` gave the following. 
```
total 54
drwxrwsr-x  2 jwsorens jemerson  6 Oct  5 11:37 .
drwxrwsr-x 18 jwsorens jemerson 45 Oct  5 11:23 ..
-rw-rw-r--  1 jwsorens jemerson  0 Sep 29 12:59 bcalm_to_gxt.log
-rw-rw-r--  1 jwsorens jemerson  0 Oct  5 11:37 cdbg.gxt
-rw-rw-r--  1 jwsorens jemerson  0 Oct  5 11:37 contigs.fa.gz
-rw-rw-r--  1 jwsorens jemerson  0 Oct  5 11:37 contigs.fa.gz.info.csv
```
If I try to run the command `python -m spacegraphcats.click search spacegraphcats/conf/RR.json --rerun-incomplete` straight from the terminal without submitting a job, in begins reading in the unitigs and the log on the screen shows...

```
...sgc is updating input file data/RR/RR452Oct_Bulk_S26_abundtrim.fq.gz with outdir prefix /share/emersonlab/sorensen/spacegraphcats
...sgc is updating input file data/RR/HQRR.18.fa with outdir prefix /share/emersonlab/sorensen/spacegraphcats
...sgc is updating input file data/RR/HQRR.25.fa with outdir prefix /share/emersonlab/sorensen/spacegraphcats
...sgc is updating input file data/RR/LQRR.84.fa with outdir prefix /share/emersonlab/sorensen/spacegraphcats
...sgc is updating input file data/RR/LQRR.8.fa with outdir prefix /share/emersonlab/sorensen/spacegraphcats
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	bcalm_catlas_input
	1	build_catlas
	1	make_contigs_kmer_index
	1	search
	4

[Mon Oct  5 11:52:01 2020]
rule bcalm_catlas_input:
    input: RR_test/bcalm.RR_test.k21.unitigs.fa
    output: RR_test_k21_r1/cdbg.gxt, RR_test_k21_r1/contigs.fa.gz, RR_test_k21_r1/contigs.fa.gz.info.csv
    jobid: 3

reading unitigs from RR_test/bcalm.RR_test.k21.unitigs.fa 
 ```
Any idea what may be causing this to happen? I'm able to run the Dory test both directly and through submitting a job.

### PR DESCRIPTION
**Pull request by ctb**
**Please note** -- this PR will CHANGE the structure of the catlas and produce different results from previous runs as a result!

Should fix #301 #296, by decreasing memory requirements!
Fixes #283, improved bcalm sorting
Fixes #299, an actual bug, and should result in increased sensitivity.

This PR addresses some of #308 by splitting the `bcalm_to_gxt` step in two --

First, `sort_bcalm_unitigs`:
* loads in the unitigs
* checks the graph structure for errors
* sorts the unitigs deterministically, into their minimum sourmash hash value.
* pickles the sorted offsets, neighbor structure, and unitig info (mean abund/size)

Then, `bcalm_to_gxt2`:
* optionally does pendant removal
* writes out the `contigs.fa.gz` file
* writes out the `cdbg.gxt` file
* writes out the `info.csv` file

The effect of this new approach is to (dramatically) reduce the memory requirements of this step, because we no longer load all the sequences into memory at once.

TODO:
- [x] fix comments at tops of files
- [x] remove bcalm_to_gxt altogether
- [x] change test_workflow
- [x] think about how to test this more properly... e.g. compare sigs with old output, etc.
- XX maybe add ksize into save files so we don't require it on command line (https://github.com/spacegraphcats/spacegraphcats/issues/320)
- [x] benchmark and compare for some real output (e.g. hu-s1)
- [x] fix CTB comments

### CODE REVIEW

**Comment by taylorreiter** [comment]
Will old catlases be back-compatible with this version? E.g. could I still run `extract_reads` on an old catlas?

**Comment by ctb** [comment]
> Will old catlases be back-compatible with this version? E.g. could I still run `extract_reads` on an old catlas?

yes! you should only run into differences if you rebuild the catlas from scratch. Search, extract, etc. should produce the same results on old catlases.

**Comment by ctb** [comment]
(excellent question, good point. have to think about how to handle versioning - this is probably a 2.0 merge!)

**Comment by ctb** [comment]
testing thoughts:
- [x] linear segment/63-only style tests for high containment
- [x] do we actually even check any numerical results for the tests? maybe add that 😆 
- [x] dig into the unitigs <-> catlas shadow checking

**Comment by ctb** [comment]
Conversation with taylor on slack:

me:
>ok I understand the problem, and have a bunch of tests in place. yay :slightly_smiling_face:

>(I had a weird heisenbug that I cannot reproduce with a clean environment and old sgc code. shrug probably from some janky bcalm stuff, Back When. I’ve written it into a test and we’ll be alerted if it happens again.)

>I’m not 100% sure what to do about merging it. I’ll sit on it for a few days. Taylor, if you point me at some of your IBD data files (raw reads, query) I can do some comparisons b/t old and new.

>but no hurry. I can tell folks suffering from memory concerns to try out the new branch w/o merging to latest.

taylor:
>ok
>raw reads: /home/tereiter/github/2020-ibd/outputs/abundtrim
>config files: /home/tereiter/github/2020-ibd/inputs/sgc_conf
>queries: /home/tereiter/github/2020-ibd/outputs/gather_matches_loso
>catlases etc: /home/tereiter/github/2020-ibd/outputs/sgc_genome_queries

**Comment by ctb** [comment]
## comparison results between old and new, running on G85166

(iHMP data set from @taylorreiter )

### unitigs are the same 👍 
```
% sourmash compare /home/tereiter/github/2020-ibd/outputs/sgc_genome_queries/G85166/*.sig G85166/*.sig

== This is sourmash version 3.5.0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

loading '/home/tereiter/github/2020-ibd/outputs/sgc_genome_queries/G85166/bcalm.<<</2020-ibd/outputs/sgc_genome_queries/G85166/bcalm.G85166.k31.unitigs.fa.sig'
loaded 1 sigs from 'G85166/bcalm.G85166.k31.unitigs.fa.sig'
loaded 2 signatures total.

0-G85166/bcalm.G8...    [1. 1.]
1-G85166/bcalm.G8...    [1. 1.]
min similarity in matrix: 1.000
```

### contigs going into the catlas are inclusive in new but missing 31% of k-mers in old ⚡ 

```
(sgc) ctbrown@farm:~/spacegraphcats$ sourmash compare /home/tereiter/github/2020-ibd/outputs/sgc_genome_queries/G85166_k31_r1/*.sig G85166_k31_r1/*.sig --containment

== This is sourmash version 3.5.0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

loading '/home/tereiter/github/2020-ibd/outputs/sgc_genome_queries/G85166_k31_r1<<</github/2020-ibd/outputs/sgc_genome_queries/G85166_k31_r1/contigs.fa.gz.sig'
loaded 1 sigs from 'G85166_k31_r1/contigs.fa.gz.sig'
loaded 2 signatures total.

0-G85166_k31_r1/c...    [1.    0.689]
1-G85166_k31_r1/c...    [1. 1.]
min similarity in matrix: 0.689
```

### query results for `RGNV35913` are ~97% inclusive in new but missing about 21% in old

```
sourmash compare --containment {/home/tereiter/github/2020-ibd/outputs/sgc_genome_queries,.}/G85166_k31_r1_search_oh0/GCF_900036035.1_RGNV35913_genomic.fna.gz.contigs.sig

== This is sourmash version 3.5.0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

loading '/home/tereiter/github/2020-ibd/outputs/sgc_genome_queries/G85166_k31_r1<<<5166_k31_r1_search_oh0/GCF_900036035.1_RGNV35913_genomic.fna.gz.contigs.sig'
loading './G85166_k31_r1_search_oh0/GCF_900036035.1_RGNV35913_genomic.fna.gz.con<<<5166_k31_r1_search_oh0/GCF_900036035.1_RGNV35913_genomic.fna.gz.contigs.sig'
loaded 2 signatures total.

0-nbhd:NZ_FCFA010...    [1.    0.787]
1-nbhd:NZ_FCFA010...    [0.97 1.  ]
min similarity in matrix: 0.787
```

### query results for `Clos_bolt_90B3` are 91% inclusive in new but missing about 40% in old.

```
(sgc) ctbrown@farm:~/spacegraphcats$ sourmash compare --containment {/home/tereiter/github/2020-ibd/outputs/sgc_genome_queries,.}/G85166_k31_r1_search_oh0/GCF_000371685.1_Clos_bolt_90B3_V1_genomic.fna.gz.contigs.sig

== This is sourmash version 3.5.0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

loading '/home/tereiter/github/2020-ibd/outputs/sgc_genome_queries/G85166_k31_r1<<<_r1_search_oh0/GCF_000371685.1_Clos_bolt_90B3_V1_genomic.fna.gz.contigs.sig'
loading './G85166_k31_r1_search_oh0/GCF_000371685.1_Clos_bolt_90B3_V1_genomic.fn<<<_r1_search_oh0/GCF_000371685.1_Clos_bolt_90B3_V1_genomic.fna.gz.contigs.sig'
loaded 2 signatures total.

0-nbhd:NZ_KB85117...    [1.    0.613]
1-nbhd:NZ_KB85117...    [0.911 1.   ]
min similarity in matrix: 0.613

**Comment by taylorreiter** [comment]
Given that I have to rerun things...should I just rebuild the catlases too? Or are those ok even though they would be different in newer versions?

**Comment by ctb** [comment]
the catlases are the problem - everything after BCALM needs to be re-run :(

**Comment by ctb** [comment]
cold take: I think we should redo the Hu analysis [published in the spacegraphcats paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02066-4) with the fixed code. The results will quite likely be much better for e.g. genome completeness and other things, but if they're not, we need to know!

If they're worse, we may face a retraction or revision. (Just to reiterate, very unlikely given what I know so far!)

If they're better, I'm not sure what to do 😂. I think ...just move forward? I'm not sure there's a need for an erratum to say "we messed up, our method is actually BETTER than we previously published." In part, because the paper was about the concept, and the math, and the implementation, and the possibilities, and the results of actually running it on real data - not **just** the results.

**Comment by ctb** [comment]
Trial re-run of hu-s1 under way on `farm:~ctbrown/2020-rerun-hu`; config etc here: https://github.com/ctb/2020-rerun-hu-s1

**Comment by ctb** [comment]
OK, I finished re-running the analysis from our genome biology paper, and did some comparisons.

## podarV containment increased quite a bit between old and new

podarV-containment

This means that, e.g., the results in figure 2 and fig 3a are probably much improved/different.

## hu checkm didn't really change, on either megahit or plass assemblies

hu-checkm

and the completeness figure (3b) is approximately the same (not shown here).

Based on this, I don't see much reason to be worried about the paper being "wrong". This fix is also, sadly, not going to lead to dramatically better biological results ;).

**Comment by ctb** [comment]
on the super positive side, though, I'm pretty solidly convinced that this code works fine!

**Comment by ctb** [comment]
Finally went ahead and benchmarked this on hu-s1, for constructing contigs.fa.gz and the gxt file.

tl;dr this PR doesn't accomplish its intended mission yet, sigh!

## this branch uses the same amount of memory and is a lot slower

stable branch: time 1198s (20m:23s); RAM: 19.6 GB
this branch: 3451s (1h:00m); RAM: 20.3 GB

## the two steps, broken down, use the same memory (!?)

this branch, sorting: time 1775s (30m:22s); RAM: 20.0 GB
this branch, output: time 1724s (30m:54s); RAM 17.2 GB

## what the heck is going on?

no idea. reading the code, I'm at a loss. further investigation needed 😆 

## Appendix A: raw data for comparison b/t branches

```
Version: 1.4.dev12+ge381790

    Command being timed: "python -m spacegraphcats.click run conf/hu-s1.json hu-s1_k31_r1/contigs.fa.gz"
    User time (seconds): 1198.61
    System time (seconds): 21.93
    Percent of CPU this job got: 99%
    Elapsed (wall clock) time (h:mm:ss or m:ss): 20:23.44
    Average shared text size (kbytes): 0
    Average unshared data size (kbytes): 0
    Average stack size (kbytes): 0
    Average total size (kbytes): 0
    Maximum resident set size (kbytes): 19587932
    Average resident set size (kbytes): 0
    Major (requiring I/O) page faults: 164
    Minor (reclaiming a frame) page faults: 7556744
    Voluntary context switches: 9294
    Involuntary context switches: 121063
    Swaps: 0
    File system inputs: 4884032
    File system outputs: 2516888
    Socket messages sent: 0
    Socket messages received: 0
    Signals delivered: 0
    Page size (bytes): 4096
    Exit status: 0
+ env


——

Version: 2.0b3.dev1+g692ac96

    Command being timed: "python -m spacegraphcats.click run conf/hu-s1.json hu-s1_k31_r1/contigs.fa.gz"
    User time (seconds): 3450.94
    System time (seconds): 174.95
    Percent of CPU this job got: 99%
    Elapsed (wall clock) time (h:mm:ss or m:ss): 1:00:42
    Average shared text size (kbytes): 0
    Average unshared data size (kbytes): 0
    Average stack size (kbytes): 0
    Average total size (kbytes): 0
    Maximum resident set size (kbytes): 20346524
    Average resident set size (kbytes): 0
    Major (requiring I/O) page faults: 165
    Minor (reclaiming a frame) page faults: 13142639
    Voluntary context switches: 17901
    Involuntary context switches: 362390
    Swaps: 0
    File system inputs: 6491520
    File system outputs: 4748032
    Socket messages sent: 0
    Socket messages received: 0
    Signals delivered: 0
    Page size (bytes): 4096
    Exit status: 0


```

## Appendix B: breakdown of two steps in this branch (sorting, then pendant removal)

```
Version: 2.0b1

Command being timed: "python -m spacegraphcats.click run conf/hu-s1.json hu-s1/bcalm.hu-s1.k31.unitigs.pickle"
    User time (seconds): 1775.05
    System time (seconds): 31.13
    Percent of CPU this job got: 99%
    Elapsed (wall clock) time (h:mm:ss or m:ss): 30:21.63
    Average shared text size (kbytes): 0
    Average unshared data size (kbytes): 0
    Average stack size (kbytes): 0
    Average total size (kbytes): 0
    Maximum resident set size (kbytes): 20346180
    Average resident set size (kbytes): 0
    Major (requiring I/O) page faults: 164
    Minor (reclaiming a frame) page faults: 6683888
    Voluntary context switches: 10039
    Involuntary context switches: 175905
    Swaps: 0
    File system inputs: 4884064
    File system outputs: 1628912
    Socket messages sent: 0
    Socket messages received: 0
    Signals delivered: 0
    Page size (bytes): 4096
    Exit status: 0


    Command being timed: "python -m spacegraphcats.click run conf/hu-s1.json hu-s1_k31_r1/contigs.fa.gz"
    User time (seconds): 1723.80
    System time (seconds): 124.89
    Percent of CPU this job got: 99%
    Elapsed (wall clock) time (h:mm:ss or m:ss): 30:53.88
    Average shared text size (kbytes): 0
    Average unshared data size (kbytes): 0
    Average stack size (kbytes): 0
    Average total size (kbytes): 0
    Maximum resident set size (kbytes): 17244056
    Average resident set size (kbytes): 0
    Major (requiring I/O) page faults: 0
    Minor (reclaiming a frame) page faults: 6837579
    Voluntary context switches: 6914
    Involuntary context switches: 182004
    Swaps: 0
    File system inputs: 1607000
    File system outputs: 3119120
    Socket messages sent: 0
    Socket messages received: 0
    Signals delivered: 0
    Page size (bytes): 4096
    Exit status: 0
```

**Comment by ctb** [comment]
hmm, one misc thought (may or may not be useful) - we could have the gxt loading code do the sorting, if we decorate the gxt file itself with the minhash for each sequence.
